### PR TITLE
fix: checkSession handles concurrent refresh token

### PR DESCRIPTION
**Pull request by rphlmr**
Issue https://github.com/mitchelvanbever/remix-auth-supabase/issues/10 describes a use case where checkSession fails because of simultaneous calls to refresh token.

___

Use case : 
```
├── invoices
│   ├── $id.js
│   └── index.js
└── invoices.js
```
If we use checkSession in $id and index loaders, both trigger refresh token.
Refresh token can be called once with the same `session.refresh_token`.

___

This PR add `pendingHandleRefreshToken` state in strategy instance.

Locally tested with a supabase JWT expiration of 10s.

@mitchelvanbever I don't know how to cover this use case in test :s if you have an idea, tell me :) 
It takes me more time trying to test than coding the fix 😂

### CODE REVIEW

**Comment by rphlmr** [comment]
Need to test https://github.com/mitchelvanbever/remix-auth-supabase/issues/10#issuecomment-1025112231 :D

**Comment by jaPajaap** [comment]
This seems to work great for me now btw. 
FYI, I tested using oauth and note that [this line](https://github.com/mitchelvanbever/remix-auth-supabase/blob/main/examples/oauth/app/routes/oauth.callback.tsx#L24) deletes the refresh token from the auth.refresh_tokens table, which invalidates that same refresh token in the session. Should I open a separate issue/PR, or would you like to fix it in the same go?

**Comment by rphlmr** [comment]
I think I can remove this, it's not part of the lib. 
Nice to see that 👍
I did that because supabase keeps a session in browser 🥲 and I didn't want that. Maybe I'm too suspicious 😅

**Comment by rphlmr** [comment]
Add an other way to remove supabase local storage session.
You may ask why :D and the reason is to make front side unaware about supabase session, outside of strategy CookieSessionStorage

**Comment by mitchelvanbever** [comment]
I've been digging around this weekend but haven't gotten around to write a test to confirm everything should be working as expected and like you also said; It takes me more time to setup environment than to actually mess around with the code.

From what I can tell the fix is solid! 

I just really want to add an accompanying test case for it, so we ensure to keep things consistent and won't break in future releases.

**Comment by rphlmr** [comment]
@jaPajaap Hello, just to warn you that there is a potential bug in the sample (onAuthStateChange listener always running, and Supabase sdk auto refresh).
I'v updated the demo ;)

**Comment by rphlmr** [comment]
@mitchelvanbever  Hey 👋 I [found a way](https://mswjs.io/docs/extensions/life-cycle-events#tracking-a-request) to test my case 🥳
What do you think about that ?

File: src/index.ts [hunk 1]
```diff
@@ -146,7 +151,13 @@ export class SupabaseStrategy extends
     const user = await this.getUser(session.access_token)
```
Comment (rphlmr) [inline src/index.ts:151]: @mitchelvanbever do you think we should to the same to reduce getUser calls or its premature optimization ?

Comment (mitchelvanbever) [reply src/index.ts:151]: I don't think we should do it. With the refreshing it results in actual failed requests whereas the getUser shouldn't error if it has a valid token. So I think if we were to apply the same pattern here as for the refreshing it could result in some weird edge cases.

**Review by mitchelvanbever — APPROVED** [review]
Thanks for finding a way to add the test. I think this should give us enough confidence to ensure it keeps working as intended!
